wfl/calculators/orca/__init__.py

```diff
@@ -0,0 +1,425 @@
+import ase.calculators.orca
+
+from wfl.calculators.utils import clean_rundir, save_results
+from wfl.pipeline import iterable_loop
+from wfl.utils.misc import atoms_to_list, chunks
+from wfl.calculators.orca.bh import BasinHoppingORCA
+
+default_keep_files = ["*.inp", "*.out", "*.ase", "*.engrad", "*.xyz",
+                        "*_trj.xyz"]
+
+
+class ORCA(ase.calculators.orca.ORCA):
+    """Extension of ASE's ORCA calculator:
+        - Specify Path to orca executable
+        - Unless specified, multiplicity is set to singlet/doublet for
+          closed-/open-shell structures
+        - Additional tasks with results stored in `extra_results` dictionary. 
+
+    Handling of "task":
+        - ORCA programs to run, in addition or instead of calculating energy forces and dipole
+        - `task` will be pre-pended to the keyword line
+        - `engrad` is enforced by default, except in the case of geometry optimisation
+        - Supported conventional tasks: `opt`, `copt` (see orca manual)
+
+    """
+
+    implemented_properties = ["energy", "forces", "dipole"]
+
+    # same as parent class, only multiplicity changed to trigger default
+    default_parameters = dict(
+        charge=0, mult=None,
+        task='engrad',
+        orcasimpleinput='tightscf PBE def2-SVP',
+        orcablocks='%scf maxiter 200 end')
+
+    def __init__(self, restart=None,
+                 ignore_bad_restart_file=Calculator._deprecated,
+                 label='orca', atoms=None,
+                 keep_files="default", dir_prefix="ORCA_", scratch_path=None,
+                 orca_path="orca", post_process=None,
+                 **kwargs):
+
+        super(ORCA, self).__init__(restart, ignore_bad_restart_file,
+                                           label, atoms, **kwargs)
+
+        self.orca_path = orca_path
+        if self.orca_path != "orca":
+            self.command = f'{orca_path} PREFIX.inp > ' \
+                           f'PREFIX.out'
+
+        self.extra_results = dict()
+        # to make use of wfl.calculators.utils.save_results()
+        self.extra_results["atoms"] = {}
+        self.extra_results["config"] = {}
+
+
+        self.keep_files = keep_files
+        # set up the directories
+        self.scratch_path = scratch_path
+        # self.directory is overwritten in self.calculate, so let's just keep track
+        self.dir_prefix = dir_prefix
+        self.wdir_base = Path(self.directory) / (self.dir_prefix + "calc_files")
+        self.wdir_base.mkdir(parents=True, exist_ok=True)
+
+        self.post_process = post_process 
+
+
+    def calculate(self, atoms=None, properties=["energy", "forces"], system_changes=all_changes):
+        """Does the calculation. Handles the working directories in addition to regular 
+        ASE calculation operations (writing input, executing, reading_results) """
+
+        Calculator.calculate(self, atoms, properties, system_changes)
+
+        if self.scratch_path is not None:
```

Comment:
The reason I didn't put it in `__init__` is because `calculate` may be called multiple times within the same calculator and I think it's more robust to do every one of them in a clean directory. For example, orca would read in the wavefunction file if it's present from before and would crash if the geometries did not match. One of the reasons for ase to re-call `calculate` is change in `Atoms` geometry.

Are you also suggesting to get rid of one of the attributes relating to directory management?